compiler/aarch64/codegen/ARM64BinaryEncoding.cpp

```diff
@@ -346,6 +347,18 @@ uint8_t *TR::ARM64Trg1ImmInstruction::generateBinaryEncoding()
    uint8_t *cursor = instructionStart;
    cursor = getOpCode().copyBinaryToBuffer(instructionStart);
    insertTargetRegister(toARM64Cursor(cursor));
+
+   if (getOpCodeValue() == TR::InstOpCode::adr)
+      {
+      TR::Symbol * sym = (getNode() && getNode()->getOpCode().hasSymbolReference()) ? getNode()->getSymbol() : NULL;
+      if (sym && sym->isStartPC())
+         {
+         intptr_t offset = reinterpret_cast<intptr_t>(reinterpret_cast<uint8_t *>(sym->getStaticSymbol()->getStaticAddress()) - cursor);
+         TR_ASSERT_FATAL((offset >= -1048576) && (offset < 1048576), "offset (%ld) is too far for adr", offset);
+         setSourceImmediate(offset & 0x1fffff); // immlo (19bit) + immhi (2bit)
+         }
+      }
+
```

Comment:
This doesn't seem like the right place for this.  For the relative distance to the startPC, I would have expected one to use a `Trg1ImmSym` instruction with the symbol reference of the startPC.  During the binary encoding of the `Trg1ImmSym` instruction it is more natural for it to alter its behaviour in the presence of a provided symbol reference for an `adr` instruction than a `Trg1Imm` instruction which (surprisingly) has to dig it out of the node.

However, the `Trg1ImmSym` instruction seems to be implemented to only expect Label symbols.  It needs to be able to handle an arbitrary symbol reference.  There doesn't appear to be any actual consumers of a `Trg1ImmSym` instruction, so altering its semantics may not be very pervasive.

If you're going to implement this, please consider using a `Trg1ImmSym` instruction.

Reply:
Okay, I will work on using `Trg1ImmSym` instruction.